tensorflow/core/kernels/data/dataset_test_base.h

```diff
@@ -0,0 +1,208 @@
+
+  // Runs an operation producing outputs. The `context` need to be initialized
+  // (see `CreateOpKernelContext()` ) before running this method.
+  Status RunOpKernel(OpKernel* op_kernel, OpKernelContext* context);
+
+  // Checks that the size of `inputs` matches the requirement of the op kernel.
+  Status CheckOpKernelInput(const OpKernel& kernel,
+                            const gtl::InlinedVector<TensorValue, 4>& inputs);
+
+  // Creates a new context for running the dataset operation.
+  Status CreateOpKernelContext(OpKernel* kernel,
+                               gtl::InlinedVector<TensorValue, 4>* inputs,
+                               std::unique_ptr<OpKernelContext>* context);
+
+  // Creates a new iterator context for iterating the dataset.
+  Status CreateIteratorContext(
+      OpKernelContext* const op_context,
+      std::unique_ptr<IteratorContext>* iterator_context);
+
+  // Creates a new serialization context for serializing the dataset and
+  // iterator.
+  Status CreateSerializationContext(
+      std::unique_ptr<SerializationContext>* context);
+
+  // Adds an arrayslice of data into the input vector. `input_types` describes
+  // the required data type for each input tensor. `shape` and `data` describes
+  // the shape and values of the current input tensor. `T` specifies the dtype
+  // of the input data.
+  template <typename T>
+  Status AddDatasetInputFromArray(gtl::InlinedVector<TensorValue, 4>* inputs,
+                                  DataTypeVector input_types,
+                                  const TensorShape& shape,
+                                  const gtl::ArraySlice<T>& data) {
+    TF_RETURN_IF_ERROR(
+        AddDatasetInput(inputs, input_types, DataTypeToEnum<T>::v(), shape));
+    test::FillValues<T>(inputs->back().tensor, data);
+    return Status::OK();
+  }
+
+ private:
+  // Adds an empty tensor with the specified dtype and shape to the input
+  // vector.
+  Status AddDatasetInput(gtl::InlinedVector<TensorValue, 4>* inputs,
+                         DataTypeVector input_types, DataType dtype,
+                         const TensorShape& shape);
+
+  // Sets the allocator attributes for the operation outputs.
+  void SetOutputAttrs();
+
+ protected:
+  std::unique_ptr<Device> device_;
+  DeviceType device_type_;
+  Allocator* allocator_;  // Owned by `AllocatorFactoryRegistry`.
+  std::vector<AllocatorAttributes> allocator_attrs_;
+  std::unique_ptr<ScopedStepContainer> step_container_;
+
+  std::unique_ptr<ProcessFunctionLibraryRuntime> pflr_;
+  FunctionLibraryRuntime* flr_;  // Owned by `pflr_`.
+  std::unique_ptr<FunctionHandleCache> function_handle_cache_;
+  std::function<void(std::function<void()>)> runner_;
+  std::unique_ptr<DeviceMgr> device_mgr_;
+  std::unique_ptr<FunctionLibraryDefinition> lib_def_;
+  std::unique_ptr<OpKernelContext::Params> params_;
+  std::unique_ptr<thread::ThreadPool> thread_pool_;
+  std::vector<Tensor*> tensors_;  // Owns Tensors.
```

Comment:
> IIUC, this field only exists to deallocated the Tensors ... why are the Tensors created by as dynamic objects instead of values (which would be deallocated when they go out of scope)?


Yeah, `tensors_ ` is used to deallocate the tensors as below. We don't want these tensors to be released when they go out of scope, because they will be used as the inputs of the op kernel in the next step.
```C++
gtl::InlinedVector<TensorValue, 4> inputs_;
...
Tensor* input = new Tensor(allocator_, dtype, shape);
tensors_.push_back(input);
inputs_->push_back({nullptr, input});
...
params_->inputs = inputs_;
...
gtl::STLDeleteElements(&tensors_);
```

Reply:
OK. Change the type of the object to `std::vector<std::unique_ptr<Tensor>>` so that you the tensors are delete automatically when the instance of this object is deleted (as opposed to having to manually delete them in the destructor).